File: Source Codes/C#/ComplexNumbers_CSharp/ComplexNumber.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ComplexNumbers_CSharp
{
    class ComplexNumber
    {
        // Members of the class
        private bool isPolar;

        private double real;
        private double imaginary;

        private double radius;
        private double angle;

        // Member calling
        double GetReal()
        {
            return real;
        }

        void SetReal(double real)
        {
            this.real = real;
        }

        double GetImaginary()
        {
            return imaginary;
        }

        void SetImaginary(double i)
        {
            imaginary = i;
        }

        double GetRadius()
        {
            if (this.isPolar)
            {
                return radius;
            }
            else
            {
                return ((this.real * this.real) + (this.imaginary * this.imaginary));
            }
        }

        void SetRadius(double r)
        {
            radius = r;
        }

        double GetAngle()
        {
            if (this.isPolar)
            {
                return angle;
            }
            else
            {
                return (Math.Atan(this.imaginary/this.real));
            }
        }

        void SetAngle(double a)
        {
            if (this.isPolar)
            {
                angle = a;
            }
        }

        // Functions
        string ToString()
        {
            string str = "";

            if (this.isPolar)
            {
                // Polar
                if (this.angle > 0)
                {
                    // Positive angle
                    str = this.radius + "(cos" + this.angle + " + isin" + this.angle + ")";
                }
                else
                {
                    str = this.radius + "(cos" + this.angle + " - isin" + -this.angle + ")";
                }
            }
            else
            {
                // Rect
                if (this.imaginary > 0)
                {
                    // Positive imaginary
                    str = this.real + " + i" + this.imaginary;
                }
                else
                {
                    str = this.real + " - i" + -this.imaginary;
                }
            }
            return str;
        }

        ComplexNumber AsPolar()
        {
            ComplexNumber number = new ComplexNumber();
            if (this.isPolar)
            {
                return this;
            }
            else
            {
                number = new ComplexNumber(true, this.GetRadius(), this.GetAngle());
            }

            return number;
        }

        ComplexNumber AsCartesian()
        {
            ComplexNumber number = new ComplexNumber();
            if (!this.isPolar)
            {
                return this;
            }
            else
            {
                number = new ComplexNumber(false, 
                    this.GetRadius() * Math.Cos(this.GetAngle()), 
                    this.GetRadius() * Math.Sin(this.GetAngle()));
            }

            return number;
        }

        ComplexNumber Logarithm()
        {
            ComplexNumber number = new ComplexNumber();
            number.real = Math.Log(this.GetRadius());
            number.imaginary = this.GetAngle();
            number.isPolar = false;
            return number;
        }

        ComplexNumber Conjugate()
        {
            ComplexNumber number = new ComplexNumber();

            if (this.isPolar)
            {
                // The number is polar
                number.radius = this.radius;
                number.angle = -this.angle;
                number.isPolar = true;
            }
            else
            {
                number.real = this.real;
                number.imaginary = -this.imaginary;
                number.isPolar = false;
            }

            return number;
        }

        ComplexNumber Reciprocal()
        {
            ComplexNumber number = new ComplexNumber();

            number.real = (this.real) / (this.real * this.real + this.imaginary * this.imaginary);
            // Minus sign shows that there is a conjugate of the complex number.
            number.imaginary = -(this.imaginary) / (this.real * this.real + this.imaginary * this.imaginary);

            number.isPolar = false;
            return number;
        }

        // Operator overloading
        static ComplexNumber operator +(ComplexNumber a, ComplexNumber b) {
            // Add then up
            if (!a.isPolar == !b.isPolar)
            {
                return new ComplexNumber(a.isPolar, a.real + b.real, a.imaginary + b.imaginary);
            }
            else
            {
                // Return exception
                throw new NotImplementedException();
            }
        }
        
        static ComplexNumber operator -(ComplexNumber a)
        {
            if (!a.isPolar)
            {
                return new ComplexNumber(a.isPolar, -a.radius, a.angle);
            }
            else
            {
                return new ComplexNumber(a.isPolar, -a.real, -a.imaginary);
            }
        }

        static ComplexNumber operator -(ComplexNumber a, ComplexNumber b)
        {
            if (!a.isPolar && !b.isPolar)
            {
                return new ComplexNumber(a.isPolar, a.real - b.real, a.imaginary - b.imaginary);
            }
            else
            {
                throw new NotImplementedException();
            }
        }

        static ComplexNumber operator *(ComplexNumber a, ComplexNumber b)
        {
            ComplexNumber number = new ComplexNumber();
            if (a.isPolar && b.isPolar)
            {
                number.radius = a.radius * b.radius;
                number.angle = a.angle + b.angle;
                number.isPolar = true;
            }
            else if (!b.isPolar && !b.isPolar)
            {
                // Both are not polar; they're rectangular
                number.real = (a.real * b.real) - (a.imaginary * b.imaginary);
                number.imaginary = (a.real * b.imaginary) + (a.imaginary * b.real);
                number.isPolar = false;
            }
            return number;
        }

        static ComplexNumber operator /(ComplexNumber a, ComplexNumber b)
        {
            ComplexNumber number = new ComplexNumber();

            if (a.isPolar && b.isPolar)
            {
                // Both are polar numbers
                number.radius = a.radius / b.radius;
                number.angle = a.angle - b.angle;
                number.isPolar = true;
            }
            else if (!a.isPolar && !b.isPolar)
            {
                // Both are rect;
                number.real = (a.real * b.real) + (a.imaginary * b.imaginary) /
                    ((b.real * b.real) + (b.imaginary * b.imaginary));

                number.imaginary = (a.real * b.real) - (a.imaginary * b.imaginary) /
                    ((b.real * b.real) + (b.imaginary * b.imaginary));
                number.isPolar = false;
            }
            return number;
        }

        static bool operator ==(ComplexNumber a, ComplexNumber b)
        {
            if (b.isPolar && a.isPolar)
            {
                // If both are polar
                if (b.angle == a.angle && b.radius == a.radius)
                {
                    return true;
                }
                else
                {
                    return false;
                }
            }
            else if (b.isPolar && !a.isPolar)
            {
                return false;
            }
            else if (!b.isPolar && a.isPolar)
            {
                return false;
            }
            else
            {
                // Check whether their values are same
                if (b.real == a.real && b.imaginary == a.imaginary)
                {
                    return true;
                }
                else
                {
                    return false;
                }
            }
        }

        // Constructors
        public ComplexNumber()
        {
            // Default
            isPolar = false;
            real = 0;
            imaginary = 0;
        }

        public ComplexNumber(bool isPolar, double firstVal, double secondVal)
        {
            // Custom 
            this.isPolar = isPolar;
            if (this.isPolar)
            {
                // Polar
                radius = firstVal;
                angle = secondVal;
            }
            else
            {
                real = firstVal;
                imaginary = secondVal;
            }
        }
    }
}

File: Source Codes/VB.NET/ComplexNumbers_VB/ComplexNumber.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace ComplexNumbers_VB
    Class ComplexNumber
        ' Members of the class
        Private isPolar As Boolean

        Private real As Double
        Private imaginary As Double

        Private radius As Double
        Private angle As Double

        ' Member calling
        Private Function GetReal() As Double
            Return real
        End Function

        Private Sub SetReal(real As Double)
            Me.real = real
        End Sub

        Private Function GetImaginary() As Double
            Return imaginary
        End Function

        Private Sub SetImaginary(i As Double)
            imaginary = i
        End Sub

        Private Function GetRadius() As Double
            If Me.isPolar Then
                Return radius
            Else
                Return ((Me.real * Me.real) + (Me.imaginary * Me.imaginary))
            End If
        End Function

        Private Sub SetRadius(r As Double)
            radius = r
        End Sub

        Private Function GetAngle() As Double
            If Me.isPolar Then
                Return angle
            Else
                Return (Math.Atan(Me.imaginary / Me.real))
            End If
        End Function

        Private Sub SetAngle(a As Double)
            If Me.isPolar Then
                angle = a
            End If
        End Sub

        ' Functions
        Private Function ToString() As String
            Dim str As String = ""

            If Me.isPolar Then
                ' Polar
                If Me.angle > 0 Then
                    ' Positive angle
                    str = Me.radius + "(cos" + Me.angle + " + isin" + Me.angle + ")"
                Else
                    str = Me.radius + "(cos" + Me.angle + " - isin" + -Me.angle + ")"
                End If
            Else
                ' Rect
                If Me.imaginary > 0 Then
                    ' Positive imaginary
                    str = Me.real + " + i" + Me.imaginary
                Else
                    str = Me.real + " - i" + -Me.imaginary
                End If
            End If
            Return str
        End Function

        Private Function AsPolar() As ComplexNumber
            Dim number As New ComplexNumber()
            If Me.isPolar Then
                Return Me
            Else
                number = New ComplexNumber(True, Me.GetRadius(), Me.GetAngle())
            End If

            Return number
        End Function

        Private Function AsCartesian() As ComplexNumber
            Dim number As New ComplexNumber()
            If Not Me.isPolar Then
                Return Me
            Else
                number = New ComplexNumber(False, Me.GetRadius() * Math.Cos(Me.GetAngle()), Me.GetRadius() * Math.Sin(Me.GetAngle()))
            End If

            Return number
        End Function

        Private Function Logarithm() As ComplexNumber
            Dim number As New ComplexNumber()
            number.real = Math.Log(Me.GetRadius())
            number.imaginary = Me.GetAngle()
            number.isPolar = False
            Return number
        End Function

        Private Function Conjugate() As ComplexNumber
            Dim number As New ComplexNumber()

            If Me.isPolar Then
                ' The number is polar
                number.radius = Me.radius
                number.angle = -Me.angle
                number.isPolar = True
            Else
                number.real = Me.real
                number.imaginary = -Me.imaginary
                number.isPolar = False
            End If

            Return number
        End Function

        Private Function Reciprocal() As ComplexNumber
            Dim number As New ComplexNumber()

            number.real = (Me.real) / (Me.real * Me.real + Me.imaginary * Me.imaginary)
            ' Minus sign shows that there is a conjugate of the complex number.
            number.imaginary = -(Me.imaginary) / (Me.real * Me.real + Me.imaginary * Me.imaginary)

            number.isPolar = False
            Return number
        End Function

        ' Operator overloading
        Shared Operator +(a As ComplexNumber, b As ComplexNumber) As ComplexNumber
            ' Add then up
            If Not a.isPolar = Not b.isPolar Then
                Return New ComplexNumber(a.isPolar, a.real + b.real, a.imaginary + b.imaginary)
            Else
                ' Return exception
                Throw New NotImplementedException()
            End If
        End Operator

        Shared Operator -(a As ComplexNumber) As ComplexNumber
            If Not a.isPolar Then
                Return New ComplexNumber(a.isPolar, -a.radius, a.angle)
            Else
                Return New ComplexNumber(a.isPolar, -a.real, -a.imaginary)
            End If
        End Operator

        Shared Operator -(a As ComplexNumber, b As ComplexNumber) As ComplexNumber
            If Not a.isPolar AndAlso Not b.isPolar Then
                Return New ComplexNumber(a.isPolar, a.real - b.real, a.imaginary - b.imaginary)
            Else
                Throw New NotImplementedException()
            End If
        End Operator

        Shared Operator *(a As ComplexNumber, b As ComplexNumber) As ComplexNumber
            Dim number As New ComplexNumber()
            If a.isPolar AndAlso b.isPolar Then
                number.radius = a.radius * b.radius
                number.angle = a.angle + b.angle
                number.isPolar = True
            ElseIf Not b.isPolar AndAlso Not b.isPolar Then
                ' Both are not polar; they're rectangular
                number.real = (a.real * b.real) - (a.imaginary * b.imaginary)
                number.imaginary = (a.real * b.imaginary) + (a.imaginary * b.real)
                number.isPolar = False
            End If
            Return number
        End Operator

        Shared Operator /(a As ComplexNumber, b As ComplexNumber) As ComplexNumber
            Dim number As New ComplexNumber()

            If a.isPolar AndAlso b.isPolar Then
                ' Both are polar numbers
                number.radius = a.radius / b.radius
                number.angle = a.angle - b.angle
                number.isPolar = True
            ElseIf Not a.isPolar AndAlso Not b.isPolar Then
                ' Both are rect;
                number.real = (a.real * b.real) + (a.imaginary * b.imaginary) / ((b.real * b.real) + (b.imaginary * b.imaginary))

                number.imaginary = (a.real * b.real) - (a.imaginary * b.imaginary) / ((b.real * b.real) + (b.imaginary * b.imaginary))
                number.isPolar = False
            End If
            Return number
        End Operator

        Shared Operator =(a As ComplexNumber, b As ComplexNumber) As Boolean
            If b.isPolar AndAlso a.isPolar Then
                ' If both are polar
                If b.angle = a.angle AndAlso b.radius = a.radius Then
                    Return True
                Else
                    Return False
                End If
            ElseIf b.isPolar AndAlso Not a.isPolar Then
                Return False
            ElseIf Not b.isPolar AndAlso a.isPolar Then
                Return False
            Else
                ' Check whether their values are same
                If b.real = a.real AndAlso b.imaginary = a.imaginary Then
                    Return True
                Else
                    Return False
                End If
            End If
        End Operator

        Shared Operator <>(a As ComplexNumber, b As ComplexNumber) As Boolean
            If b.isPolar AndAlso a.isPolar Then
                ' If both are polar
                If b.angle = a.angle AndAlso b.radius = a.radius Then
                    Return True
                Else
                    Return False
                End If
            ElseIf b.isPolar AndAlso Not a.isPolar Then
                Return False
            ElseIf Not b.isPolar AndAlso a.isPolar Then
                Return False
            Else
                ' Check whether their values are same
                If b.real = a.real AndAlso b.imaginary = a.imaginary Then
                    Return True
                Else
                    Return False
                End If
            End If
        End Operator

        ' Constructors
        Public Sub New()
            ' Default
            isPolar = False
            real = 0
            imaginary = 0
        End Sub

        Public Sub New(isPolar As Boolean, firstVal As Double, secondVal As Double)
            ' Custom 
            Me.isPolar = isPolar
            If Me.isPolar Then
                ' Polar
                radius = firstVal
                angle = secondVal
            Else
                real = firstVal
                imaginary = secondVal
            End If
        End Sub
    End Class
End Namespace
